File: backend/src/main/java/com/nhtl/models/Produit.java
package com.nhtl.models;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "produits")
public class Produit {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false)
    private ServiceType serviceType;

    @Column(nullable = false)
    private String nom;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal prix;

    private String devise = "EUR";

    private String categorie;

    @Column(name = "image_url", columnDefinition = "TEXT")
    private String imageUrl;

    @Convert(converter = StringListConverter.class)
    @Column(name = "image_urls", columnDefinition = "TEXT")
    private List<String> imageUrls = new ArrayList<>();

    @Column(nullable = false)
    private Integer stock = 0;

    private String unite;  // kg, unité, boîte, litre…

    @Column(nullable = false)
    private Boolean actif = true;

    @Column(name = "date_ajout")
    private LocalDateTime dateAjout;

    @Column(name = "date_modification")
    private LocalDateTime dateModification;

    @PrePersist
    protected void onCreate() {
        dateAjout        = LocalDateTime.now();
        dateModification = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        dateModification = LocalDateTime.now();
    }

    // ── Getters & Setters ──────────────────────────────────────────────────

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public ServiceType getServiceType() { return serviceType; }
    public void setServiceType(ServiceType serviceType) { this.serviceType = serviceType; }

    public String getNom() { return nom; }
    public void setNom(String nom) { this.nom = nom; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public BigDecimal getPrix() { return prix; }
    public void setPrix(BigDecimal prix) { this.prix = prix; }

    public String getDevise() { return devise; }
    public void setDevise(String devise) { this.devise = devise; }

    public String getCategorie() { return categorie; }
    public void setCategorie(String categorie) { this.categorie = categorie; }

    public String getImageUrl() { return imageUrl; }
    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }

    public List<String> getImageUrls() { return imageUrls; }
    public void setImageUrls(List<String> imageUrls) { this.imageUrls = imageUrls != null ? imageUrls : new ArrayList<>(); }

    public Integer getStock() { return stock; }
    public void setStock(Integer stock) { this.stock = stock; }

    public String getUnite() { return unite; }
    public void setUnite(String unite) { this.unite = unite; }

    public Boolean getActif() { return actif; }
    public void setActif(Boolean actif) { this.actif = actif; }

    public LocalDateTime getDateAjout() { return dateAjout; }
    public void setDateAjout(LocalDateTime dateAjout) { this.dateAjout = dateAjout; }

    public LocalDateTime getDateModification() { return dateModification; }
    public void setDateModification(LocalDateTime dateModification) { this.dateModification = dateModification; }
}

File: frontend/lib/models/produit.dart
// lib/models/produit.dart

class Produit {
  final int? id;
  final String serviceType;
  final String nom;
  final String? description;
  final double prix;
  final String devise;
  final String? categorie;
  final String? imageUrl;
  final List<String> imageUrls;
  final int stock;
  final String? unite;
  final bool actif;
  final DateTime? dateAjout;
  final DateTime? dateModification;

  bool get enStock => stock > 0;

  Produit({
    this.id,
    required this.serviceType,
    required this.nom,
    this.description,
    required this.prix,
    this.devise = 'EUR',
    this.categorie,
    this.imageUrl,
    this.imageUrls = const [],
    this.stock = 0,
    this.unite,
    this.actif = true,
    this.dateAjout,
    this.dateModification,
  });

  factory Produit.fromJson(Map<String, dynamic> json) {
    final rawUrls = json['imageUrls'];
    final List<String> imageUrls = rawUrls is List
        ? rawUrls.map((e) => e.toString()).toList()
        : [];
    return Produit(
      id: (json['id'] as num?)?.toInt(),
      serviceType: json['serviceType'] as String? ?? '',
      nom: json['nom'] as String? ?? '',
      description: json['description'] as String?,
      prix: (json['prix'] is int)
          ? (json['prix'] as int).toDouble()
          : (json['prix'] as double? ?? 0.0),
      devise: json['devise'] as String? ?? 'EUR',
      categorie: json['categorie'] as String?,
      imageUrl: json['imageUrl'] as String?,
      imageUrls: imageUrls,
      stock: (json['stock'] as num?)?.toInt() ?? 0,
      unite: json['unite'] as String?,
      actif: json['actif'] as bool? ?? true,
      dateAjout: json['dateAjout'] != null
          ? DateTime.tryParse(json['dateAjout'].toString())
          : null,
      dateModification: json['dateModification'] != null
          ? DateTime.tryParse(json['dateModification'].toString())
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'serviceType': serviceType,
        'nom': nom,
        'description': description,
        'prix': prix,
        'devise': devise,
        'categorie': categorie,
        'imageUrl': imageUrl,
        'imageUrls': imageUrls,
        'stock': stock,
        'unite': unite,
        'actif': actif,
      };
}
